Bug: grad(jit(custom_linear_solve)) encounters ad_util.Zero in backwards pass

I'm not quite sure what specifically is going wrong yet, but here's an example traceback from https://github.com/google/jax/pull/1535:

    jax/test_util.py:179: in check_grads
        _check_grads(f, args, order)
    jax/test_util.py:177: in _check_grads
        _check_grads(f_vjp, args, order - 1)
    jax/test_util.py:172: in _check_grads
        _check_vjp(f, partial(api.vjp, f), args)
    jax/test_util.py:148: in check_vjp
        cotangent_out = conj(vjpfun(conj(cotangent)))
    jax/api_util.py:66: in apply_flat_fun_nokwargs
        ans = fun(*args)
    jax/interpreters/ad.py:114: in vjp_
        _, arg_cts = backward_pass(jaxpr, consts, (), dummy_args, dummy_primals_and_cts)
    jax/interpreters/ad.py:178: in backward_pass
        eqn.params, subjaxpr, sub_consts, sub_freevar_vals, invals, cts_in)
    jax/interpreters/ad.py:458: in call_transpose
        out_flat = primitive.bind(fun, *all_args, **params)
    jax/core.py:591: in call_bind
        outs = primitive.impl(f, *args, **params)
    jax/interpreters/xla.py:418: in _xla_call_impl
        compiled_fun = _xla_callable(fun, device, backend, *map(abstractify, args))
    jax/linear_util.py:217: in cached_fun
        ans, f_prev = cached_fun_body(f, args)
    jax/linear_util.py:214: in cached_fun_body
        return call(f, *args), f
    jax/interpreters/xla.py:434: in _xla_callable
        jaxpr, (pvals, consts, env) = pe.trace_to_subjaxpr(fun, master, False).call_wrapped(pvals)
    jax/linear_util.py:165: in call_wrapped
        ans = self.f(*args, **dict(self.params, **kwargs))
    jax/interpreters/ad.py:178: in backward_pass
        eqn.params, subjaxpr, sub_consts, sub_freevar_vals, invals, cts_in)
    jax/interpreters/ad.py:458: in call_transpose
        out_flat = primitive.bind(fun, *all_args, **params)
    jax/core.py:594: in call_bind
        outs = map(full_lower, top_trace.process_call(primitive, f, tracers, params))
    jax/interpreters/partial_eval.py:111: in process_call
        out_flat = call_primitive.bind(fun, *in_consts, **params)
    jax/core.py:591: in call_bind
        outs = primitive.impl(f, *args, **params)
    jax/interpreters/xla.py:418: in _xla_call_impl
        compiled_fun = _xla_callable(fun, device, backend, *map(abstractify, args))
    jax/linear_util.py:217: in cached_fun
        ans, f_prev = cached_fun_body(f, args)
    jax/linear_util.py:214: in cached_fun_body
        return call(f, *args), f
    jax/interpreters/xla.py:434: in _xla_callable
        jaxpr, (pvals, consts, env) = pe.trace_to_subjaxpr(fun, master, False).call_wrapped(pvals)
    jax/linear_util.py:165: in call_wrapped
        ans = self.f(*args, **dict(self.params, **kwargs))
    jax/interpreters/ad.py:181: in backward_pass
        cts_out = get_primitive_transpose(eqn.primitive)(cts_in, *invals, **eqn.params)
    jax/lax_linalg.py:358: in triangular_solve_transpose_rule
        not transpose_a, conjugate_a, unit_diagonal)
    jax/lax_linalg.py:80: in triangular_solve
        conjugate_a=conjugate_a, unit_diagonal=unit_diagonal)
    jax/core.py:155: in bind
        out_tracer = top_trace.process_primitive(self, tracers, kwargs)
    jax/interpreters/partial_eval.py:91: in process_primitive
        tracers = map(self.instantiate_const, tracers)
    jax/util.py:39: in safe_map
        return list(map(f, *args))
    jax/interpreters/partial_eval.py:80: in instantiate_const
        return self.new_instantiated_const(const)
    jax/interpreters/partial_eval.py:66: in new_instantiated_const
        return JaxprTracer(self, PartialVal((get_aval(val), unit)), ConstVar(val))
    jax/core.py:538: in get_aval
        return concrete_aval(x)
    _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

    x = Zero

        def concrete_aval(x):
          try:
            return pytype_aval_mappings[type(x)](x)
          except KeyError:
    >       raise TypeError("{} is not a valid Jax type".format(type(x)))
    E       TypeError: <class 'jax.ad_util.Zero'> is not a valid Jax type

--- a/jax/lax_linalg.py
+++ b/jax/lax_linalg.py
@@ -354,8 +354,11 @@ def triangular_solve_transpose_rule(
   # Triangular solve is nonlinear in its first argument and linear in its second
   # argument, analogous to `div` but swapped.
   assert a is not ad.undefined_primal and b is ad.undefined_primal
-  cotangent_b = triangular_solve(a, cotangent, left_side, lower,
-                                 not transpose_a, conjugate_a, unit_diagonal)
+  if cotangent is ad_util.zero:
+    cotangent_b = ad_util.zero
+  else:
+    cotangent_b = triangular_solve(a, cotangent, left_side, lower,
+                                   not transpose_a, conjugate_a, unit_diagonal)
   return [None, cotangent_b]
 
 

--- a/tests/lax_control_flow_test.py
+++ b/tests/lax_control_flow_test.py
@@ -1227,6 +1227,9 @@ class LaxControlFlowTest(jtu.JaxTestCase):
 
     jtu.check_grads(linear_solve, (a, b), order=2)
 
+    # regression test for https://github.com/google/jax/issues/1536
+    jtu.check_grads(api.jit(linear_solve), (a, b), order=2)
+
   def test_custom_linear_solve_without_transpose_solve(self):
 
     def explicit_jacobian_solve(matvec, b):

